builder/rewrite.h

```diff
@@ -0,0 +1,428 @@
+#ifndef SLINKY_BUILDER_REWRITE_H
+#define SLINKY_BUILDER_REWRITE_H
+
+#include "builder/substitute.h"
+#include "runtime/evaluate.h"
+#include "runtime/expr.h"
+
+// This pattern matching engine is heavily inspired by https://github.com/halide/Halide/blob/main/src/IRMatch.h.
+
+namespace slinky {
+namespace rewrite {
+
+constexpr int symbol_count = 4;
+constexpr int constant_count = 3;
+
+struct match_context {
+  const base_expr_node* vars[symbol_count];
+  const index_t* constants[constant_count];
+  int variant;
+  int variant_bits;
+
+  void clear() {
+    // Memset initializing these makes debug builds significantly faster.
```

Comment:
Interesting! Any idea why?

Reply:
Also: should we also zero out `variant` and `variant_bits`? I'd expect all of the struct to be predictably cleared after this call.